File: wiki/src/main/java/io/github/freya022/wiki/java/autocomplete/transformer/FullNameTransformer.java
package io.github.freya022.wiki.java.autocomplete.transformer;

import io.github.freya022.botcommands.api.commands.application.slash.autocomplete.AutocompleteTransformer;
import io.github.freya022.botcommands.api.core.service.annotations.BService;
import io.github.freya022.wiki.switches.wiki.WikiLanguage;
import net.dv8tion.jda.api.interactions.commands.Command;
import org.jetbrains.annotations.NotNull;

@WikiLanguage(WikiLanguage.Language.JAVA)
// --8<-- [start:autocomplete_transformer-java]
@BService
public class FullNameTransformer implements AutocompleteTransformer<FullName> {
    @Override
    public @NotNull Class<FullName> getElementType() {
        return FullName.class;
    }

    @NotNull
    @Override
    public Command.Choice apply(@NotNull FullName fullName) {
        return new Command.Choice(
                "%s %s".formatted(fullName.firstName(), fullName.secondName()),
                "%s|%s".formatted(fullName, fullName.secondName())
        );
    }
}
// --8<-- [end:autocomplete_transformer-java]

File: wiki/src/main/kotlin/io/github/freya022/wiki/autocomplete/transformer/FullNameTransformer.kt
package io.github.freya022.wiki.autocomplete.transformer

import io.github.freya022.botcommands.api.commands.application.slash.autocomplete.AutocompleteTransformer
import io.github.freya022.botcommands.api.core.service.annotations.BService
import io.github.freya022.wiki.switches.wiki.WikiLanguage
import net.dv8tion.jda.api.interactions.commands.Command

@WikiLanguage(WikiLanguage.Language.KOTLIN)
// --8<-- [start:full_name_obj-kotlin]
data class FullName(val firstName: String, val secondName: String)
// --8<-- [end:full_name_obj-kotlin]

@WikiLanguage(WikiLanguage.Language.KOTLIN)
// --8<-- [start:autocomplete_transformer-kotlin]
@BService
class FullNameTransformer : AutocompleteTransformer<FullName> {
    override fun getElementType(): Class<FullName> = FullName::class.java

    override fun apply(e: FullName): Command.Choice {
        return Command.Choice("${e.firstName} ${e.secondName}", "${e.firstName}|${e.secondName}")
    }
}
// --8<-- [end:autocomplete_transformer-kotlin]